### ISSUE TITLE
zrepl intermittent socket connection errors

### ISSUE THREAD

**Issue by johnramsden** [zrepl/zrepl#262]
## Summary

I've been having intermittent slowdowns of the daemon on the pull side, it seems to occur after the daemon has been running for several hours. It is noticeable in the status page where errors flash about the Unix socket and `EOF`, or trailer.

Once this starts occurring there seems to be a significant slowdown, and after restarting the demon normal performance is seen again.

## System

**Server:**

* OS: FreeNAS-11.2-U7
* `server: zrepl version=v0.2.0 GOOS=freebsd GOARCH=amd64 Compiler=gc`

**Client:**

* Arch Linux 5.4.10-arch1-1
* `client: zrepl version=arch-0.2.1 GOOS=linux GOARCH=amd64 Compiler=gc`

## Configuration

Here's my configuration on the pull server side, I have multiple hosts I'm pulling from but their configuration is all identical.

```
global:
  logging:

    - type: "stdout"
      level:  "info"
      format: "human"

jobs:
  - name: wooly
    type: pull
    connect:
      type: tls
      address: "172.20.20.3:8888"
      ca: "/mnt/tank/system/root/etc/zrepl/wooly.ramsden.network.crt"
      cert: "/mnt/tank/system/root/etc/zrepl/lilan.ramsden.network.crt"
      key: "/mnt/tank/system/root/etc/zrepl/lilan.ramsden.network.key"
      server_cn: "wooly.ramsden.network"
    root_fs: "tank/repl/wooly"
    interval: 15m
    pruning:
      keep_sender:
        - type: grid
          grid: 1x1h(keep=all) | 24x1h | 30x1d
          regex: "zrepl_.*"
      keep_receiver:
        - type: grid
          grid: 1x1h(keep=all) | 24x1h | 60x1d | 24x30d
          regex: "zrepl_.*"
```

The client servers all have multiple datasets, around 18+ that are being replicated.

**Comment by problame** [comment]
What exactly is slowing down? just the status reporting or the entire replication?
Could you observe whether the slow-down is due to CPU usage?
If so, please try to capture a [FlameGraph](http://www.brendangregg.com/flamegraphs.html) next time the CPU usage is high, and post it here.

**Comment by johnramsden** [comment]
It's the entire application. Things seem to stutter between 0 bytes and a few b/s when this occurs.

I will watch out for next time it occurs and check my CPU usage.

**Comment by problame** [comment]
If it's not CPU-bound, maybe try to capture some pprof traces. Are you familiar with Go pprof and the `zrepl pprof` subcommand?

**Comment by johnramsden** [comment]
I am not, but I can try. Do I need a go compiler or will `zrepl` be enough?

**Comment by problame** [comment]
A good introduction: https://jvns.ca/blog/2017/09/24/profiling-go-with-pprof/
No, you don't need the go compiler, but no matter whether you run `go tool pprof` on your laptop or FreeNAS directly, you need to pass it the zrepl binary that is executing, i.e., the one on FreeNAS.

**Comment by problame** [comment]
Also: https://github.com/zrepl/zrepl/wiki/Debugging

**Comment by johnramsden** [comment]
So it had planned on waiting until this happened again but unfortunately I think I fixed the problem temporarily.

I had over 3,000 snapshots they were out of sync, and this was occurring as they tried to sync up, I set a job to restart the service once a day before this bug was reported, not realizing it might have something to do with a large number of snapshots. 

With the restart, my snapshots were able to get back in sync, and now I haven't had the bug occur again for several days even without restarting the service.

What would your recommendation be? Can I gather any more information for you, or do I need to get my snapshots out of sync?

**Comment by problame** [comment]
> With the restart, my snapshots were able to get back in sync, and now I haven't had the bug occur again for several days even without restarting the service.

Just to be clear: 3000 snapshots or 3000 filesystems?

**Comment by johnramsden** [comment]
Snapshots. 18+ datasets, each with ~3000

**Comment by problame** [comment]
> Things seem to stutter between 0 bytes and a few b/s when this occurs.

The rate displayed in the status view is not reliable if there are errors inbetween because it is computed in the client (very hacky, I know). The status response from the server contains an monotonically increasing counter. If possible, use the `zrepl status --raw` flag to extract the raw values. The `jq` tool could be helpful.
But unless we can reproduce the situation the above isn't helpful.

> What would your recommendation be? Can I gather any more information for you, or do I need to get my snapshots out of sync?

I think you can create a dummy dataset and restart the source-side zrepl to make the snapshotter stay in the sync up phase until the other filesystems would be regularly snapshotted.
Then observe replication progress.
(Although I doubt it's the snapshotter that is causing the performance problem.)

**Comment by problame** [comment]
Sorry this took so long. I pushed a commit (should show up above this comment) that enables configuration of read and write timeouts via environment variables. Could you try the CI binaries on your system & play around with the environment variables on the daemon where you observe the `EOF`s?

The patch is based on the upcoming 0.3 release, which I invite you to test (you will have to build the docs yourself or download them from CI, too (`zrepl-noarch.tar`).

**Comment by johnramsden** [comment]
Sure. I haven't actually experienced the error in quite a while, I guess because my snapshots caught up. I will create a few thousand and test.

**Comment by johnramsden** [comment]
So, I created a few thousand snapshots each with a little bit of changes to a file between snapshots, (appending a number) to try to recreate the issue with replication, but they just replicated fairly quickly with no issues.

Currently trying to replicate with multiple datasets to see if the problem will occur:

```bash
for d in {0..14}; do
    for s in {0..2500}; do
         echo "test$d - $s" | tee -a /mnt/t$d/test.log
         zfs snapshot vault/data/test$d@$s;
    done &
done
```

Once/if this reproduces the error I will try again with the new version to see if it occurs.

**Comment by johnramsden** [comment]
So @problame, I wasn't able to reproduce the error with all those snapshots. Should I give it another try before I try the new version?

**Comment by problame** [comment]
We are experiencing some problems with stale holds ( see #280  #277 for symptoms of this ). Seems to be a problem with pull jobs mostly.
So, in order to avoid problems on your systems, you should maybe stay on the 0.2, make sure that you don't restart the daemon (remember your workaround!) and see whether the problem returns.

**Comment by johnramsden** [comment]
I'll do that, let me know if you have a way I can be of more help diagnosing issue.

**Comment by johnramsden** [comment]
@problame this issue is occurring for me again. Replication seems incredibly slow, stuck at 0 B/s most of the time, and occasionally replicating. Around every 30 seconds I get a post `http://unix/status: EOF`

Right now there are 7 datasets, each with only seven snapshots to replicate, they're completing but it incredibly slowly. There's a total of 238MiB of data to replicate between all the snapshots.

Additionally one of my datasets shows:

```
stream: read error: read tcp 172.20.30.2:11943->172.20.20.2:8888: read: connection reset by peer
```

I'm not sure if that's relevant.

How should I proceed to get you more information? I'm afraid if I try using the new version for 0.3 the restart will just remove the issue for now.

**Comment by problame** [comment]
Use `zrepl pprof on 127.0.0.1:someport`, then use pprof to find out why things are slow: https://blog.golang.org/pprof

I can help via Voice call if you want, please shoot me an email or join #zrepl on freenode.

**Comment by problame** [comment]
Summary of the voice call:

* the zrepl daemon is being run on FreeNAS using the `zrepl daemon` utility
* FreeBSD top and ports htop indicate high cpu usage by zrepl daemon (150-260%) and by the parent daemon(8) tool (99.9%)
* => investigated CPU performance using go pprof and flamegraphs
* flamegraphs show that only ~8-9% of total CPU time go to zrepl
* pprof also indicates very little CPU time spent (10s sampling, only 610ms of on-cpu time observed)
  * almost all spent in the go runtime, in particular some mutex implementation

(all of the above was _without_ `zrepl status` running, didn't measure how things are with `zrepl status` running)

* ran truss on the zrepl daemon, mostly inconclusive, most of the time spent on mutex aquisition
* the ongoing replication was going very slowly, i.e. lots of "nothing happens" periods between replication steps
* once we restarted the daemon, things went back to normal
* amount of snapshots / datasets / etc unsuspicious

going forward:

* john keeps running 0.2.1, wait and see when this happens again
* no insight into why the daemon was slow
* will probably revisit once #307 lands, although since it won't be backported to 0.2.1, the new observations will also show possible slow downs by other changes in upcoming 0.3 (see #310 )

**Comment by problame** [comment]
@johnramsden I just published https://github.com/zrepl/zrepl/releases/tag/v0.3.0-rc1 which includes the activity tracing patch. And we need testers for v0.3.0-rc1. I suggest you upgrade your system to the release-candidate and we'll see whether this issue is still present in 0.3.

**Comment by johnramsden** [comment]
@problame I'll give that a try and let you know what I find.

**Comment by johnramsden** [comment]
@problame Sorry I took so long to start testing the candidate. I'm now running the release candidate on two machines, with a FreeBSD jail pull.

A little more information - the error also would occur in a FreeBSD (FreeNAS) jail not just when running on the host. I'm running with the default configuration from the FreeBSD package. I saw very high CPU usage and the connection errors in the FreeBSD jail.

To test the release candidate I I'm running the same service as the FreeBSD package provides but with the different binary.

My two clients are on Arch Linux, And have been upgraded to the release candidate.

I'll let you know if I noticed the error with the release candidate.

**Comment by johnramsden** [comment]
I'm not getting the Unix socket connection error but I am getting the extremely slow performance. See [video](https://asciinema.org/a/BRWpd1atP9hVXX9wcaxRDTG8P)

If I have time on the weekend I will set up Prometheus.

**Comment by johnramsden** [comment]
After restart it's a [bit better](https://asciinema.org/a/pYKy3nuH9Dha1y6ZT2fCh54rw):

**Comment by johnramsden** [comment]
However it does still periodically have socket connection errors. It is less frequent than on 0.2.0.

**Comment by problame** [comment]
Let's use the 0.3 activity tracing feature to debug this:

Run

```
zrepl pprof on 127.0.0.1:12345
zrepl pprof activity-trace 127.0.0.1:12345 > zrepl-activity-trace.json
```

**on both sender and receiver** **while the replication is running**.
Do that 3 times

* once for approx 30s
* once for approx 2min 
* once for approx 10min

(Stop the capture using Ctrl-C).

The activity trace shouldn't contain information beyond dataset and snapshot names, maybe have a look at them though before posting them here.

You can open the `.json` files in Chrome under `chrome://tracing` btw.
Read this doc comment for more info:

https://github.com/zrepl/zrepl/blob/master/daemon/logging/trace/trace.go#L1-L91

**Comment by johnramsden** [comment]
I got the following:

```
zrepl-v0.3.0rc1 pprof listen 127.0.0.1:12346
connecting to zrepl daemon
finished
```

```
zrepl-v0.3.0rc1 pprof activity-trace 127.0.0.1:12346 > zrepl-activity-trace.json
attaching to activity trace stream ws://127.0.0.1:12346/debug/zrepl/activity-trace
error: websocket.Dial ws://127.0.0.1:12346/debug/zrepl/activity-trace: dial tcp 127.0.0.1:12346: connect: can't assign requested address
exiting after error
```

**Comment by problame** [comment]
Is something else already listening on that port? I suppose it's pretty self-explanatory that these commands open a local TCP port.

**Comment by johnramsden** [comment]
No, and I tried multiple ports, 1234, 12345, 12346, 12349. All have the same behaviour.

```
netstat -a -n 
Active Internet connections (including servers)
Proto Recv-Q Send-Q Local Address          Foreign Address        (state)
tcp4       0      0 172.20.30.44.15608     172.20.20.3.8888       CLOSE_WAIT
tcp4       0      0 172.20.30.44.50488     172.20.20.2.8888       CLOSE_WAIT
tcp4       0      0 172.20.30.44.22737     172.20.20.2.8888       ESTABLISHED
tcp4       0      0 172.20.30.44.60356     172.20.20.3.8888       ESTABLISHED
Active UNIX domain sockets
Address          Type   Recv-Q Send-Q            Inode             Conn             Refs          Nextref Addr
fffff8003afd32d0 stream      0      0 fffff80174f54000                0                0                0 /var/run/zrepl/control
```

**Comment by problame** [comment]
If you are running in a jail, maybe bind to the Jail's IP?

**Comment by johnramsden** [comment]
Yes I am, I was wondering if that might be the issue but binding to the jails IP doesn't help.

**Comment by problame** [comment]
Huh, what happens when you restart the daemon and try the command immediately after, before things start to hang?

**Comment by johnramsden** [comment]
Same thing.

Now that replication is caught up it's not hanging anyway, it only seems to occur when he gets behind. I should have time tomorrow or sunday to setup Prometheus. Will that give you the relevant information as well?

I can then let replication get behind to cause the behavior.

**Comment by problame** [comment]
I'll try to replicate your specific situation with 17 datasets and 450 snapshots per dataset. There are some FIXMEs in the code where we have O(n^2) complexity...

**Comment by problame** [comment]
So I created a replication scenario with 17 datasets with 450 snapshot each. On my Ryzen 1700x, inside a (Debian!) VM, I observed approximately 15% CPU usage. That should rule out any concerns with regards to computational complexity (O(n^2))

The activity trace on server and client will be essential for me to understand where the daemon spends its time.
If `zrepl pprof listen on 127.0.0.1:12345` doesn't work (we'll figure out why, but that's a separate issue for now), you can start `zrepl daemon` with env var `ZREPL_ACTIVITY_TRACE=/tmp/zrepl.activity.json`. The daemon will write the activity trace to that file. Note that the file can grow quite large (tens of megabyte per minute on a busy system).
Also, the tooling to analyze the trace doesn't cope well with large traces (> 200 MBs or long time spans), so I suggest that once you observe that the daemon is hanging, you restart the daemons on both sides with that env var being set.

Please upload both side's tracing to GitHub.
You can open them yourself in Chrome on `chrome://tracing` -> `Load` button.

**Comment by johnramsden** [comment]
I'm creating a large number of snapshots and I will do as you mentioned.

Problem is restarting it seems to help, so I'm not sure if that will be very helpful.

What I will do is wait until it slows, restart and then get it on FreeBSD and linux.

If I can help you debug the trace issue I'm happy to give you a remote session again.

To clarify, will Prometheus help at all?

**Comment by problame** [comment]
Prometheus _might_ be useful to gather general system stats, but not zrepl stats. So ... I wouldn't waste my time on standing up a Prometheus setup just to debug this.

**Comment by problame** [comment]
When creating your test scenario, you will need to `sleep 1.1` between taking snapshots, see https://github.com/zrepl/zrepl/issues/346 .

**Comment by problame** [comment]
idempotent script to create the test ansphsots:

```
#!/usr/bin/env bash
set -euo pipefail

fsroot="$1"
fscount=8
snapperfscount=100

ignoreexists () {
    if [ $? = 0 ]; then
        return 0
    fi
    if [[ "$1" =~ *dataset*exists* ]];then
        return 0
    fi
    echo "error: $1"
    return 1
}


set +e
ignoreexists "$(zfs create -o mountpoint=none "$fsroot")"
ignoreexists "$(zfs snapshot "$fsroot@initial")"
set -e

for i in $(seq 1 "$fscount"); do
    set +e
    ignoreexists "$(zfs create -o mountpoint=none "$fsroot/fs$i")"
    ignoreexists "$(zfs snapshot "$fsroot/fs$i@initial")"
    set -e
done

echo -n 'do initial replication, then hit enter'
read

for j in $(seq 1 "$snapperfscount"); do
    for i in $(seq 1 "$fscount"); do
        (
        set +e
        if ignoreexists "$(zfs snapshot "$fsroot/fs$i@$j")"; then
            sleep 1.2
        fi
        set -e
        )&
    done
    for i in $(seq 1 "$fscount"); do
        wait
    done
done
```

**Comment by johnramsden** [comment]
So I ran it it for several hours with the activity monitor (it didn't get above a hundred megabytes), I can't reproduce the issue. Part of the issue is that happens periodically, and even less often with 0.3.0.

I could try again and leave it for a longer and see if I can reproduce, restarting it seems to fix the issue for a while, so I can't really just start it up and then restart it when I see the issue.

**Comment by problame** [comment]
Any news  / change in behavior while running the 0.3 release candidates?

**Comment by johnramsden** [comment]
It occurs occasionally with the RC. I can't seem to run the tracing once the service has started though, since I can't connect to the port.

**Comment by problame** [comment]
Have you tried starting the daemon with `ZREPL_ACTIVITY_TRACE=/tmp/zrepl.activity.json` then? Watch the file size of the trace, though.

**Comment by johnramsden** [comment]
Yes, but unfortunately it takes a few days for the problem to occur, is it okay to leave it running that long?

**Comment by problame** [comment]
In general: yes. Make sure you have sufficient disk space for the .json (watch one replication interval to gauge how much space it takes on your system).

**Comment by 3nprob** [comment]
Seeing this now - seems consistent with observed behavior above. Since the resolution of #463  , there is now a backlog of 4000-9000 snapshots over 5 filesystems. Pull jobs, all on 0.5.0.

In zrepl status I see `status fetch: Post "http://unix/status": EOF` and `status fetch: http: unexpected EOF reading trailer`.

zrepl log is continuously logging `control handler io error err="write unix /var/run/zrepl/control->@: i/o timeout"` with 2s intervals on receiver.

The receiving host is a bit RAM constrained, but has plenty of headroom in swap, which is barely used (`vm.swappiness=40`).

Replication is also extremely slow - at least an order of magnitude less than expected - in the single kb/s , though somehow progressing. Will see if I can manage to get some tracing out of it.

It does start occuring shortly after startup, so replicating is not an issue.

**Comment by 3nprob** [comment]
So preliminarily, increasing timeouts to 10s and the above does not occur so far. Still mighty slow, but not at all as terrible as it was previously. Will have to attach some additional physical storage to safely do the activity trace and profiling.

**Comment by problame** [comment]
>  Still mighty slow, but not at all as terrible as it was previously. 

What is slow? Replication itself, or the status reporting? How long does one `zrepl status --raw` take on average?

**Comment by problame** [comment]
(Just to avoid false positives, you are running 0.5.0 + [the patch](https://github.com/zrepl/zrepl/issues/379#issuecomment-991889521), right?)
